scapy/contrib/ikev2.py

```diff
@@ -723,11 +740,40 @@ class IKEv2_Nonce(IKEv2_Payload):
 class IKEv2_Notify(IKEv2_Payload):
     name = "IKEv2 Notify"
     fields_desc = IKEv2_Payload.fields_desc[:3] + [
-        ByteEnumField("proto", None, {0: "Reserved", 1: "IKE", 2: "AH", 3: "ESP"}),  # noqa: E501
+        ByteEnumField("proto", None, IKEv2ProtocolTypes),
         FieldLenField("SPIsize", None, "SPI", "B"),
         ShortEnumField("type", 0, IKEv2NotifyMessageTypes),
         XStrLenField("SPI", "", length_from=lambda pkt: pkt.SPIsize),
-        XStrLenField("notify", "", length_from=lambda pkt: pkt.length - 8),
+        ConditionalField(
+            XStrLenField("notify", "", length_from=lambda pkt: pkt.length - 8),
+            lambda pkt: pkt.type not in (16407, 16408)
+        ),
+        ConditionalField(
+            # REDIRECT, REDIRECTED_FROM  (RFC 5685)
+            ByteEnumField("gw_id_type", 1, IKEv2GatewayIDTypes),
+            lambda pkt: pkt.type in (16407, 16408)
+        ),
+        ConditionalField(
+            # REDIRECT, REDIRECTED_FROM  (RFC 5685)
+            FieldLenField("gw_id_len", None, "gw_id", "B"),
+            lambda pkt: pkt.type in (16407, 16408)
+        ),
+        ConditionalField(
+            # REDIRECT, REDIRECTED_FROM  (RFC 5685)
+            MultipleTypeField(
+                [
+                    (IPField("gw_id", "127.0.0.1"), lambda x: x.gw_id_type == 1),
+                    (IP6Field("gw_id", "::1"), lambda x: x.gw_id_type == 5),
+                ],
+                StrLenField("gw_id", "", length_from=lambda x: x.gw_id_len)
+            ),
+            lambda pkt: pkt.type in (16407, 16408)
+        ),
+        ConditionalField(
```

Comment:
@gpotter2 is there a more elegant way to implement this 'polymorphism' of the `notify` payload?  In the generic case it should just be an `XStrLenField`, but in special cases, a more specific dissection is possible.

```
XStrLenField("notify", "", length_from=lambda pkt: pkt.length - 8),
```

In other words: is there something like a `UnionField`, where every case of the union can contain one or more fields?

Reply:
There are more notification types with a more specific `notify` payload. Adding them all in this fashion will become very messy.

Reply:
> special cases

Then can't you use a MultipleTypeField ?

Reply:
A MultipleTypeField only works if the fields have the same name = same role, but different types, which is mostly what a union is. If there are optional fields with different meanings, use `ConditionalField`s

Reply:
Ok, since I have have a field (`gw_id`) which is both conditional and of a union type, my construction seems not so bad then. The only thing which still bothers me is that I have four `ConditionalField`s which check exactly the same condition `lambda pkt: pkt.type in (16407, 16408)`. Is it possible to have a single `ConditionalField` which contains all four fields?

Reply:
Can't you just combine them then?
`ConditionalField(MultipleTypeField(<...>), <condition>)`

Reply:
That's what I already did, see for example

https://github.com/secdev/scapy/blob/eb9ba2d7579b0cce0b7d6f661098342806aa5b24/scapy/contrib/ikev2.py#L761-L771

My question was aiming in a different direction: I have three `ConditionalField`s  (`gw_id_type`, `gw_id_len`, and `gw_id`) which actually belong together and are guarded by the same condition, namely that it's a  `REDIRECT` or `REDIRECTED_FROM` notification. I was just wandering whether it was possible to have a single `ConditionalField` (or a similar class) which combines all three fields instead of having three separate `ConditionalField`s.

The reason why I asked is because many more `ConditionalField`s might be added in the future for other notification types and the code would remain better readable if there would be only one `ConditionalField` per notification type.

But never mind my question, it seems like it was more confusing than helpful. Anyway, the pull request in its current form works as desired and has tests, i.e., it's ready for final review.